File: python/examples/tool_router/preload.py
"""
Tool Router preload with OpenAI Agents.

Shows direct tool exposure for:
1. Composio tools via preload["tools"].
2. SDK custom tools via preload=True on the custom tool or toolkit.
3. A nested custom tool override with preload=False inside a preloaded toolkit.

Usage:
    COMPOSIO_API_KEY=... OPENAI_API_KEY=... python examples/tool_router/preload.py
"""

import os

from agents import Agent, Runner
from composio_openai_agents import OpenAIAgentsProvider
from pydantic import BaseModel, Field

from composio import Composio


def require_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"Set {name} before running this example.")
    return value


class UserLookupInput(BaseModel):
    user_id: str = Field(description="Internal user ID, for example user-1")


class TeamSearchInput(BaseModel):
    team: str = Field(description="Team name to search for")


class AccountInput(BaseModel):
    user_id: str = Field(description="Internal user ID")


INTERNAL_USERS = {
    "user-1": {
        "id": "user-1",
        "name": "Ada Lovelace",
        "email": "ada@example.com",
        "team": "platform",
        "plan": "enterprise",
    },
    "user-2": {
        "id": "user-2",
        "name": "Grace Hopper",
        "email": "grace@example.com",
        "team": "developer-tools",
        "plan": "startup",
    },
}


composio = Composio(
    api_key=require_env("COMPOSIO_API_KEY"),
    base_url=os.environ.get("COMPOSIO_BASE_URL"),
    provider=OpenAIAgentsProvider(),
)
require_env("OPENAI_API_KEY")


@composio.experimental.tool(
    slug="LOOKUP_INTERNAL_USER",
    name="Lookup internal user",
    description="Look up an internal demo user profile by user ID.",
    preload=True,
)
def lookup_internal_user(input: UserLookupInput, ctx):
    user = INTERNAL_USERS.get(input.user_id)
    if not user:
        raise ValueError(f'User "{input.user_id}" not found')
    return user


@composio.experimental.tool(
    slug="SEARCH_INTERNAL_USERS",
    name="Search internal users",
    description=(
        "Search demo internal users by team. This custom tool is search-only "
        "because preload is not enabled."
    ),
)
def search_internal_users(input: TeamSearchInput, ctx):
    return {
        "results": [
            user for user in INTERNAL_USERS.values() if user["team"] == input.team
        ]
    }


internal_admin = composio.experimental.Toolkit(
    slug="INTERNAL_ADMIN",
    name="Internal admin",
    description="Demo internal administration tools.",
    preload=True,
)


@internal_admin.tool(
    slug="GET_ACCOUNT_HEALTH",
    name="Get account health",
    description="Return internal account health details for a user.",
)
def get_account_health(input: AccountInput, ctx):
    return {
        "user_id": input.user_id,
        "health": "green",
        "open_incidents": 0,
        "renewal_risk": "low",
    }


@internal_admin.tool(
    slug="GET_ACCOUNT_AUDIT_LOG",
    name="Get account audit log",
    description=(
        "Return internal account audit events. This overrides the toolkit "
        "preload default and remains search-only."
    ),
    preload=False,
)
def get_account_audit_log(input: AccountInput, ctx):
    return {
        "user_id": input.user_id,
        "events": ["login", "settings_viewed"],
    }


session = composio.create(
    user_id="preload-example-user",
    toolkits=["hackernews"],
    tools={"hackernews": {"enable": ["HACKERNEWS_GET_USER"]}},
    preload={"tools": ["HACKERNEWS_GET_USER"]},
    manage_connections=False,
    experimental={
        "custom_tools": [lookup_internal_user, search_internal_users],
        "custom_toolkits": [internal_admin],
    },
)

tools = session.tools()
tool_names = [tool.name for tool in tools]
assert "HACKERNEWS_GET_USER" in tool_names
assert "LOCAL_LOOKUP_INTERNAL_USER" in tool_names
assert "LOCAL_INTERNAL_ADMIN_GET_ACCOUNT_HEALTH" in tool_names
assert "LOCAL_SEARCH_INTERNAL_USERS" not in tool_names
assert "LOCAL_INTERNAL_ADMIN_GET_ACCOUNT_AUDIT_LOG" not in tool_names

print("Direct tools exposed to the agent:")
for tool in tools:
    print(f"- {tool.name}")

agent = Agent(
    name="Preload Demo Agent",
    instructions="Use the provided tools to perform the task.",
    model=os.environ.get("OPENAI_MODEL", "gpt-5.5"),
    tools=tools,
)

prompt = (
    'Look up Hacker News user "pg", then look up internal user user-1 and '
    "their account health. Summarize the useful facts."
)
result = Runner.run_sync(agent, prompt)
print(result.final_output)

File: ts/examples/tool-router/src/preload.ts
/**
 * Tool Router preload with OpenAI Agents
 *
 * Shows direct tool exposure for:
 * 1. Composio tools via `preload.tools`
 * 2. SDK custom tools via `preload: true` on the custom tool or toolkit
 *
 * Usage:
 *   COMPOSIO_API_KEY=... OPENAI_API_KEY=... bun src/preload.ts
 */
import 'dotenv/config';
import assert from 'node:assert/strict';
import { Composio, experimental_createTool, experimental_createToolkit } from '@composio/core';
import { OpenAIAgentsProvider } from '@composio/openai-agents';
import { Agent, run } from '@openai/agents';
import { z } from 'zod/v3';

function requireEnv(name: 'COMPOSIO_API_KEY' | 'OPENAI_API_KEY'): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Set ${name} before running this example.`);
  }
  return value;
}

const composioApiKey = requireEnv('COMPOSIO_API_KEY');
requireEnv('OPENAI_API_KEY');

const internalUsers: Record<string, Record<string, unknown>> = {
  'user-1': {
    id: 'user-1',
    name: 'Ada Lovelace',
    email: 'ada@example.com',
    team: 'platform',
    plan: 'enterprise',
  },
  'user-2': {
    id: 'user-2',
    name: 'Grace Hopper',
    email: 'grace@example.com',
    team: 'developer-tools',
    plan: 'startup',
  },
};

const lookupInternalUser = experimental_createTool('LOOKUP_INTERNAL_USER', {
  name: 'Lookup internal user',
  description: 'Look up an internal demo user profile by user ID.',
  preload: true,
  inputParams: z.object({
    user_id: z.string().describe('Internal user ID, for example user-1'),
  }),
  execute: async ({ user_id }) => {
    const user = internalUsers[user_id];
    if (!user) {
      throw new Error(`User "${user_id}" not found`);
    }
    return user;
  },
});

const searchInternalUsers = experimental_createTool('SEARCH_INTERNAL_USERS', {
  name: 'Search internal users',
  description:
    'Search demo internal users by team. This custom tool is search-only because preload is not enabled.',
  inputParams: z.object({
    team: z.string().describe('Team name to search for'),
  }),
  execute: async ({ team }) => ({
    results: Object.values(internalUsers).filter(user => user.team === team),
  }),
});

const getAccountHealth = experimental_createTool('GET_ACCOUNT_HEALTH', {
  name: 'Get account health',
  description: 'Return internal account health details for a user.',
  inputParams: z.object({
    user_id: z.string().describe('Internal user ID'),
  }),
  execute: async ({ user_id }) => ({
    user_id,
    health: 'green',
    open_incidents: 0,
    renewal_risk: 'low',
  }),
});

const getAccountAuditLog = experimental_createTool('GET_ACCOUNT_AUDIT_LOG', {
  name: 'Get account audit log',
  description:
    'Return internal account audit events. This overrides the toolkit preload default and remains search-only.',
  preload: false,
  inputParams: z.object({
    user_id: z.string().describe('Internal user ID'),
  }),
  execute: async ({ user_id }) => ({
    user_id,
    events: ['login', 'settings_viewed'],
  }),
});

const internalAdmin = experimental_createToolkit('INTERNAL_ADMIN', {
  name: 'Internal admin',
  description: 'Demo internal administration tools.',
  preload: true,
  tools: [getAccountHealth, getAccountAuditLog],
});

const composio = new Composio({
  apiKey: composioApiKey,
  baseURL: process.env.COMPOSIO_BASE_URL || undefined,
  provider: new OpenAIAgentsProvider(),
});

const session = await composio.create('preload-example-user', {
  toolkits: ['hackernews'],
  tools: {
    hackernews: {
      enable: ['HACKERNEWS_GET_USER'],
    },
  },
  preload: {
    tools: ['HACKERNEWS_GET_USER'],
  },
  manageConnections: false,
  experimental: {
    customTools: [lookupInternalUser, searchInternalUsers],
    customToolkits: [internalAdmin],
  },
});

const tools = await session.tools();
const toolNames = tools.map(tool => tool.name);
assert(toolNames.includes('HACKERNEWS_GET_USER'));
assert(toolNames.includes('LOCAL_LOOKUP_INTERNAL_USER'));
assert(toolNames.includes('LOCAL_INTERNAL_ADMIN_GET_ACCOUNT_HEALTH'));
assert(!toolNames.includes('LOCAL_SEARCH_INTERNAL_USERS'));
assert(!toolNames.includes('LOCAL_INTERNAL_ADMIN_GET_ACCOUNT_AUDIT_LOG'));

console.log('Direct tools exposed to the agent:');
for (const tool of tools) {
  console.log(`- ${tool.name}`);
}

const agent = new Agent({
  name: 'Preload Demo Agent',
  instructions: 'Use the provided tools to perform the task.',
  model: process.env.OPENAI_MODEL ?? 'gpt-5.5',
  tools,
});

const prompt =
  process.argv.slice(2).join(' ') ||
  'Look up Hacker News user "pg", then look up internal user user-1 and their account health. Summarize the useful facts.';

console.log(`\nUser: ${prompt}\n`);
const result = await run(agent, prompt);
console.log(`Assistant: ${result.finalOutput}`);
